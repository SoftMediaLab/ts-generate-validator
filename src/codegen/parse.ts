import { parseStruct, Decorator, ImportNode } from 'ts-file-parser';
import * as fs from 'fs';
// import * as path from 'path';
// import { ArrayType, BasicType } from 'ts-file-parser';

export interface DecoratedClassFieldMetadata {
  name: string;
  type: any;
  optional: any;
  decorators: Decorator[];
}

export interface DecoratedClassMetadata {
  name: string;
  decorators: Decorator[];
  fields: DecoratedClassFieldMetadata[];
}

export interface InputFileMetadata {
  name: string;
  imports: string[];
  classes: DecoratedClassMetadata[];
}

export const parseInputFiles = (files: string[]): InputFileMetadata[] => {
  return files.map((file) => {
    console.log(`Reading ${file}...`);
    const content = fs.readFileSync(file, 'utf-8');
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const structure = parseStruct(content, {}, file);

    //if (file.match(/^.*common.ts$/)) {
    //console.log(JSON.stringify(structure, null, 4));
    //}

    return {
      name: structure.name.replace(/\/+/g, '.'),
      imports: buildImports(structure._imports),
      classes: []
    };
  });
};

const buildImports = (importNodes: ImportNode[]): string[] => {
  return importNodes.map((iNode) => {
    return `import { ${iNode.clauses.join(', ')} } from ${iNode.absPathNode.join('/')}`;
  });
};
