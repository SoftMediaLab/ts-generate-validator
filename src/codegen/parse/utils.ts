import { primitiveValidationTypes } from './../../validators/model';
import { ValidationType } from '../../validators/model';

export const getValidationTypeByTypeName = (typeName: string): ValidationType => {
  if (primitiveValidationTypes.includes(typeName as any)) {
    return <ValidationType>typeName;
  }

  const notSupportedTypes = ['null', 'undefined', 'symbol', 'function', 'object'];
  if (notSupportedTypes.includes(typeName)) {
    return ValidationType.notSupported;
  }

  return ValidationType.unknown;
};
