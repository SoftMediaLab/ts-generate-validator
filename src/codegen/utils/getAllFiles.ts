import * as fs from 'fs';

export const getAllFiles = (path: string): string[] => {
  const tsRegExp = /.+\.ts$/;
  const returnFiles: string[] = [];

  const files = fs.readdirSync(path);
  files.forEach((file) => {
    const endPath = `${path}/${file}`;
    if (fs.statSync(endPath).isDirectory()) {
      const subFiles = getAllFiles(endPath);
      returnFiles.push(...subFiles);
    } else {
      const matches = tsRegExp.exec(endPath);
      if (matches && matches.length > 0) {
        returnFiles.push(matches[0]);
      }
    }
  });
  return returnFiles;
};
