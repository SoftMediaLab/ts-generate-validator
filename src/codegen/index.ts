import { parseInputFiles } from './parse';
import { getAllFiles } from './utils/getAllFiles';
import { getCodegenConfig } from './../config/codegen';

export const createValidators = (): void => {
  const config = getCodegenConfig();
  const inputFiles = getAllFiles(config.inputPath);
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const inputFilesMetadata = parseInputFiles(inputFiles);

  // TODO: implement
};
