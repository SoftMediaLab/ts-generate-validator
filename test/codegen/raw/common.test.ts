import { parseInputFiles } from '../../../src/codegen/parse/index';
import * as path from 'path';
import { getAllFiles } from '../../../src/codegen/utils/getAllFiles';
import { getCodegenConfig } from '../../../src/config/codegen';
import { deleteConfigFile, createConfigFile } from '../../utils/config';

describe('codegen/raw/common', () => {
  const prevRootDir = process.cwd();

  beforeAll(() => {
    const testRootDir = process.cwd() + '/test/mock/raw/common';
    process.chdir(testRootDir);
    createConfigFile({
      inputPath: '.'
    });
  });

  afterAll(() => {
    deleteConfigFile();
    process.chdir(prevRootDir);
  });

  test('parse metadata', () => {
    const config = getCodegenConfig();
    const inputFiles = getAllFiles(path.resolve(config.inputPath));
    const inputFilesMetadata = parseInputFiles(inputFiles);

    expect(inputFilesMetadata).toMatchSnapshot('parseMetadata');
  });
});
