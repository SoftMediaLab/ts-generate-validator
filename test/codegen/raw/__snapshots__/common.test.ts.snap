// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen/raw/common create validators: generated validators at "src/generated/validators/model.ts" 1`] = `
"/* This file was automatically generated and should not be edited */
// tslint:disable
/* eslint-disable */

import {
  GeneratedValidation,
  ValidationConfig,
  ValidationError,
  ValidationException,
  Data,
  UserContext,
  getConfig,
  mergeDeep,
  typeValidator,
  customValidator,
  equalValidator,
  equalToValidator
} from 'ts-generate-validator';

import {
  TypeValidatorWithCustomMessage,
  TypeValidatorOnWrongPropertyType,
  TypeValidatorOnNullPropertyType,
  CustomValidatorFailed,
  CutomValidatorSuccess,
  CutomValidatorOnNonPrimitiveStructure,
  CutomValidatorSuccessAsync,
  IgnoreValidationForNonPrimitiveProperty,
  IgnoreValidationForPrimitiveProperty,
  IgnoreValidationForPrimitivePropertyWithOtherDecorators,
  EqualValidatorWithDefaultMessage,
  EqualValidatorWithCustomMessage,
  EqualValidatorForBoolean,
  EqualToValidatorWithDefaultMessage
} from './../../../model';

import { customValidatorFailedValidator } from './model';

export const typeValidatorWithCustomMessageValidator: GeneratedValidation = <
  D extends Data = typeof TypeValidatorWithCustomMessage,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
    typeValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      customMessage: 'type custom message',

      type: 'string'
    });
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const typeValidatorOnWrongPropertyTypeValidator: GeneratedValidation = <
  D extends Data = typeof TypeValidatorOnWrongPropertyType,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
    typeValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      customMessage: 'type custom message',

      typeDescription: customValidatorFailedValidator,

      type: 'nested'
    });
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const typeValidatorOnNullPropertyTypeValidator: GeneratedValidation = <
  D extends Data = typeof TypeValidatorOnNullPropertyType,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
    typeValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      type: 'null'
    });
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const customValidatorFailedValidator: GeneratedValidation = <
  D extends Data = typeof CustomValidatorFailed,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
    typeValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      type: 'number'
    });

    customValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      customValidationFunction: () => {
        throw new Error('Failed!');
      }
    });
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const cutomValidatorSuccessValidator: GeneratedValidation = <
  D extends Data = typeof CutomValidatorSuccess,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
    typeValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      type: 'number'
    });

    customValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      customValidationFunction: () => undefined
    });
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const cutomValidatorOnNonPrimitiveStructureValidator: GeneratedValidation = <
  D extends Data = typeof CutomValidatorOnNonPrimitiveStructure,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
    customValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      customValidationFunction: () => undefined
    });
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const cutomValidatorSuccessAsyncValidator: GeneratedValidation = async <
  D extends Data = typeof CutomValidatorSuccessAsync,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
    typeValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      type: 'number'
    });

    await customValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      customValidationFunction: async (): Promise<void> => new Promise((resolve) => setTimeout(() => resolve(), 300))
    });
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const ignoreValidationForNonPrimitivePropertyValidator: GeneratedValidation = <
  D extends Data = typeof IgnoreValidationForNonPrimitiveProperty,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const ignoreValidationForPrimitivePropertyValidator: GeneratedValidation = <
  D extends Data = typeof IgnoreValidationForPrimitiveProperty,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const ignoreValidationForPrimitivePropertyWithOtherDecoratorsValidator: GeneratedValidation = <
  D extends Data = typeof IgnoreValidationForPrimitivePropertyWithOtherDecorators,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const equalValidatorWithDefaultMessageValidator: GeneratedValidation = <
  D extends Data = typeof EqualValidatorWithDefaultMessage,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
    typeValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      type: 'string'
    });

    equalValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      value: 'abcdef'
    });
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const equalValidatorWithCustomMessageValidator: GeneratedValidation = <
  D extends Data = typeof EqualValidatorWithCustomMessage,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
    typeValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      type: 'number'
    });

    equalValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      value: 245,

      customMessage: 'equal custom message'
    });
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const equalValidatorForBooleanValidator: GeneratedValidation = <
  D extends Data = typeof EqualValidatorForBoolean,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
    typeValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      type: 'boolean'
    });

    equalValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      value: true
    });
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};

export const equalToValidatorWithDefaultMessageValidator: GeneratedValidation = <
  D extends Data = typeof EqualToValidatorWithDefaultMessage,
  C extends UserContext = UserContext
>(
  data: D,
  configArg?: ValidationConfig,
  context?: C
) => {
  const config = <ValidationConfig>mergeDeep({}, getConfig(), configArg ?? {});
  const errors: ValidationError[] = [];

  try {
    typeValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      type: 'string'
    });

    equalToValidator({
      property: data?.someProperty,
      propertyName: 'someProperty',
      data,
      config,
      context,

      targetPropertyName: 'someOtherProperty'
    });

    typeValidator({
      property: data?.someOtherProperty,
      propertyName: 'someOtherProperty',
      data,
      config,
      context,

      type: 'string'
    });

    equalToValidator({
      property: data?.someOtherProperty,
      propertyName: 'someOtherProperty',
      data,
      config,
      context,

      targetPropertyName: 'someProperty'
    });
  } catch (err) {
    if (!(err instanceof ValidationError)) {
      throw err;
    }

    errors.push(err);

    if (config.stopAtFirstError) {
      throw new ValidationException(errors);
    }
  }

  if (errors.length) {
    throw new ValidationException(errors);
  }
};
"
`;

exports[`codegen/raw/common parse & prepare: data for render 1`] = `
Array [
  Object {
    "fileName": "model.ts",
    "filePath": "src/generated/validators",
    "imports": Array [
      Object {
        "clauses": "GeneratedValidation, ValidationConfig, ValidationError, ValidationException, Data, UserContext, getConfig, mergeDeep, typeValidator, customValidator, equalValidator, equalToValidator",
        "path": "ts-generate-validator",
      },
      Object {
        "clauses": "TypeValidatorWithCustomMessage, TypeValidatorOnWrongPropertyType, TypeValidatorOnNullPropertyType, CustomValidatorFailed, CutomValidatorSuccess, CutomValidatorOnNonPrimitiveStructure, CutomValidatorSuccessAsync, IgnoreValidationForNonPrimitiveProperty, IgnoreValidationForPrimitiveProperty, IgnoreValidationForPrimitivePropertyWithOtherDecorators, EqualValidatorWithDefaultMessage, EqualValidatorWithCustomMessage, EqualValidatorForBoolean, EqualToValidatorWithDefaultMessage",
        "path": "./../../../model",
      },
      Object {
        "clauses": "customValidatorFailedValidator",
        "path": "./model",
      },
    ],
    "validations": Array [
      Object {
        "async": false,
        "items": Array [
          Object {
            "propertyName": "someProperty",
            "validatorName": "typeValidator",
            "validatorPayload": Array [
              Object {
                "property": "customMessage",
                "type": "string",
                "value": "type custom message",
              },
              Object {
                "property": "type",
                "type": "string",
                "value": "string",
              },
            ],
          },
        ],
        "modelClassName": "TypeValidatorWithCustomMessage",
        "name": "typeValidatorWithCustomMessageValidator",
      },
      Object {
        "async": false,
        "items": Array [
          Object {
            "propertyName": "someProperty",
            "validatorName": "typeValidator",
            "validatorPayload": Array [
              Object {
                "property": "customMessage",
                "type": "string",
                "value": "type custom message",
              },
              Object {
                "property": "typeDescription",
                "type": "object",
                "value": "customValidatorFailedValidator",
              },
              Object {
                "property": "type",
                "type": "string",
                "value": "nested",
              },
            ],
          },
        ],
        "modelClassName": "TypeValidatorOnWrongPropertyType",
        "name": "typeValidatorOnWrongPropertyTypeValidator",
      },
      Object {
        "async": false,
        "items": Array [
          Object {
            "propertyName": "someProperty",
            "validatorName": "typeValidator",
            "validatorPayload": Array [
              Object {
                "property": "type",
                "type": "string",
                "value": "null",
              },
            ],
          },
        ],
        "modelClassName": "TypeValidatorOnNullPropertyType",
        "name": "typeValidatorOnNullPropertyTypeValidator",
      },
      Object {
        "async": false,
        "items": Array [
          Object {
            "propertyName": "someProperty",
            "validatorName": "typeValidator",
            "validatorPayload": Array [
              Object {
                "property": "type",
                "type": "string",
                "value": "number",
              },
            ],
          },
          Object {
            "async": false,
            "propertyName": "someProperty",
            "validatorName": "customValidator",
            "validatorPayload": Array [
              Object {
                "property": "customValidationFunction",
                "type": "",
                "value": "() => {
    throw new Error('Failed!');
  }",
              },
            ],
          },
        ],
        "modelClassName": "CustomValidatorFailed",
        "name": "customValidatorFailedValidator",
      },
      Object {
        "async": false,
        "items": Array [
          Object {
            "propertyName": "someProperty",
            "validatorName": "typeValidator",
            "validatorPayload": Array [
              Object {
                "property": "type",
                "type": "string",
                "value": "number",
              },
            ],
          },
          Object {
            "async": false,
            "propertyName": "someProperty",
            "validatorName": "customValidator",
            "validatorPayload": Array [
              Object {
                "property": "customValidationFunction",
                "type": "",
                "value": "() => undefined",
              },
            ],
          },
        ],
        "modelClassName": "CutomValidatorSuccess",
        "name": "cutomValidatorSuccessValidator",
      },
      Object {
        "async": false,
        "items": Array [
          Object {
            "async": false,
            "propertyName": "someProperty",
            "validatorName": "customValidator",
            "validatorPayload": Array [
              Object {
                "property": "customValidationFunction",
                "type": "",
                "value": "() => undefined",
              },
            ],
          },
        ],
        "modelClassName": "CutomValidatorOnNonPrimitiveStructure",
        "name": "cutomValidatorOnNonPrimitiveStructureValidator",
      },
      Object {
        "async": true,
        "items": Array [
          Object {
            "propertyName": "someProperty",
            "validatorName": "typeValidator",
            "validatorPayload": Array [
              Object {
                "property": "type",
                "type": "string",
                "value": "number",
              },
            ],
          },
          Object {
            "async": true,
            "propertyName": "someProperty",
            "validatorName": "customValidator",
            "validatorPayload": Array [
              Object {
                "property": "customValidationFunction",
                "type": "",
                "value": "async (): Promise<void> => new Promise((resolve) => setTimeout(() => resolve(), 300))",
              },
            ],
          },
        ],
        "modelClassName": "CutomValidatorSuccessAsync",
        "name": "cutomValidatorSuccessAsyncValidator",
      },
      Object {
        "async": false,
        "items": Array [],
        "modelClassName": "IgnoreValidationForNonPrimitiveProperty",
        "name": "ignoreValidationForNonPrimitivePropertyValidator",
      },
      Object {
        "async": false,
        "items": Array [],
        "modelClassName": "IgnoreValidationForPrimitiveProperty",
        "name": "ignoreValidationForPrimitivePropertyValidator",
      },
      Object {
        "async": false,
        "items": Array [],
        "modelClassName": "IgnoreValidationForPrimitivePropertyWithOtherDecorators",
        "name": "ignoreValidationForPrimitivePropertyWithOtherDecoratorsValidator",
      },
      Object {
        "async": false,
        "items": Array [
          Object {
            "propertyName": "someProperty",
            "validatorName": "typeValidator",
            "validatorPayload": Array [
              Object {
                "property": "type",
                "type": "string",
                "value": "string",
              },
            ],
          },
          Object {
            "async": false,
            "propertyName": "someProperty",
            "validatorName": "equalValidator",
            "validatorPayload": Array [
              Object {
                "property": "value",
                "type": "string",
                "value": "abcdef",
              },
            ],
          },
        ],
        "modelClassName": "EqualValidatorWithDefaultMessage",
        "name": "equalValidatorWithDefaultMessageValidator",
      },
      Object {
        "async": false,
        "items": Array [
          Object {
            "propertyName": "someProperty",
            "validatorName": "typeValidator",
            "validatorPayload": Array [
              Object {
                "property": "type",
                "type": "string",
                "value": "number",
              },
            ],
          },
          Object {
            "async": false,
            "propertyName": "someProperty",
            "validatorName": "equalValidator",
            "validatorPayload": Array [
              Object {
                "property": "value",
                "type": "number",
                "value": "245",
              },
              Object {
                "property": "customMessage",
                "type": "string",
                "value": "equal custom message",
              },
            ],
          },
        ],
        "modelClassName": "EqualValidatorWithCustomMessage",
        "name": "equalValidatorWithCustomMessageValidator",
      },
      Object {
        "async": false,
        "items": Array [
          Object {
            "propertyName": "someProperty",
            "validatorName": "typeValidator",
            "validatorPayload": Array [
              Object {
                "property": "type",
                "type": "string",
                "value": "boolean",
              },
            ],
          },
          Object {
            "async": false,
            "propertyName": "someProperty",
            "validatorName": "equalValidator",
            "validatorPayload": Array [
              Object {
                "property": "value",
                "type": "boolean",
                "value": "true",
              },
            ],
          },
        ],
        "modelClassName": "EqualValidatorForBoolean",
        "name": "equalValidatorForBooleanValidator",
      },
      Object {
        "async": false,
        "items": Array [
          Object {
            "propertyName": "someProperty",
            "validatorName": "typeValidator",
            "validatorPayload": Array [
              Object {
                "property": "type",
                "type": "string",
                "value": "string",
              },
            ],
          },
          Object {
            "async": false,
            "propertyName": "someProperty",
            "validatorName": "equalToValidator",
            "validatorPayload": Array [
              Object {
                "property": "targetPropertyName",
                "type": "string",
                "value": "someOtherProperty",
              },
            ],
          },
          Object {
            "propertyName": "someOtherProperty",
            "validatorName": "typeValidator",
            "validatorPayload": Array [
              Object {
                "property": "type",
                "type": "string",
                "value": "string",
              },
            ],
          },
          Object {
            "async": false,
            "propertyName": "someOtherProperty",
            "validatorName": "equalToValidator",
            "validatorPayload": Array [
              Object {
                "property": "targetPropertyName",
                "type": "string",
                "value": "someProperty",
              },
            ],
          },
        ],
        "modelClassName": "EqualToValidatorWithDefaultMessage",
        "name": "equalToValidatorWithDefaultMessageValidator",
      },
    ],
  },
]
`;

exports[`codegen/raw/common parse & prepare: parsed metadata 1`] = `
Array [
  Object {
    "classes": Array [
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "type custom message",
                ],
                "name": "TypeValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
        ],
        "name": "TypeValidatorWithCustomMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "type custom message",
                ],
                "name": "TypeValidation",
              },
            ],
            "name": "someProperty",
            "optional": true,
            "type": Object {
              "name": "CustomValidatorFailed",
              "referencePath": "model.ts",
              "validationType": "nested",
            },
          },
        ],
        "name": "TypeValidatorOnWrongPropertyType",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "null",
              "validationType": "null",
            },
          },
        ],
        "name": "TypeValidatorOnNullPropertyType",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "() => {
    throw new Error('Failed!');
  }",
                ],
                "name": "CustomValidation",
              },
            ],
            "name": "someProperty",
            "optional": true,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "CustomValidatorFailed",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "() => undefined",
                ],
                "name": "CustomValidation",
              },
            ],
            "name": "someProperty",
            "optional": true,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "CutomValidatorSuccess",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "() => undefined",
                ],
                "name": "CustomValidation",
              },
            ],
            "name": "someProperty",
            "optional": true,
            "type": Object {
              "name": "mock",
              "referencePath": "model.ts",
              "validationType": "notSupported",
            },
          },
        ],
        "name": "CutomValidatorOnNonPrimitiveStructure",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "async (): Promise<void> => new Promise((resolve) => setTimeout(() => resolve(), 300))",
                ],
                "name": "CustomValidation",
              },
            ],
            "name": "someProperty",
            "optional": true,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "CutomValidatorSuccessAsync",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "IgnoreValidation",
              },
            ],
            "name": "someProperty",
            "optional": true,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "IgnoreValidationForNonPrimitiveProperty",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "IgnoreValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "IgnoreValidationForPrimitiveProperty",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  10,
                ],
                "name": "MaxValidation",
              },
              Object {
                "arguments": Array [],
                "name": "IgnoreValidation",
              },
              Object {
                "arguments": Array [
                  2,
                ],
                "name": "MinValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "IgnoreValidationForPrimitivePropertyWithOtherDecorators",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "abcdef",
                ],
                "name": "EqualValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
        ],
        "name": "EqualValidatorWithDefaultMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  245,
                  "equal custom message",
                ],
                "name": "EqualValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
        ],
        "name": "EqualValidatorWithCustomMessage",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  true,
                ],
                "name": "EqualValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "boolean",
              "validationType": "boolean",
            },
          },
        ],
        "name": "EqualValidatorForBoolean",
      },
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "someOtherProperty",
                ],
                "name": "EqualToValidation",
              },
            ],
            "name": "someProperty",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "someProperty",
                ],
                "name": "EqualToValidation",
              },
            ],
            "name": "someOtherProperty",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
        ],
        "name": "EqualToValidatorWithDefaultMessage",
      },
    ],
    "name": "model.ts",
  },
]
`;
